src/nekrs_driver.cpp

```diff
@@ -0,0 +1,229 @@
+  Position c{0., 0., 0.};
+  double mass = 0.;
+  for (gsl::index i = 0; i < n_gll_; ++i) {
+    auto idx = local_elem * n_gll_ + i;
+    c.x += x_[idx] * mass_matrix_[idx];
+    c.y += y_[idx] * mass_matrix_[idx];
+    c.z += z_[idx] * mass_matrix_[idx];
+    mass += mass_matrix_[idx];
+  }
+  c.x /= mass;
+  c.y /= mass;
+  c.z /= mass;
+  return c;
+}
+
+std::vector<Position> NekRSDriver::centroid_local() const
+{
+  std::vector<Position> c(n_local_elem());
+  for (int32_t i = 0; i < n_local_elem(); ++i) {
+    c[i] = this->centroid_at(i);
+  }
+  return c;
+}
+
+double NekRSDriver::volume_at(int32_t local_elem) const
+{
+  Expects(local_elem < n_local_elem());
+  double v = 0.;
+  for (int32_t i = 0; i < n_gll_; ++i) {
+    v += mass_matrix_[local_elem * n_gll_ + i];
+  }
+  return v;
+}
+
+std::vector<double> NekRSDriver::volume_local() const
+{
+  std::vector<double> v(n_local_elem());
+  for (int32_t i = 0; i < n_local_elem(); ++i) {
+    v[i] = this->volume_at(i);
+  }
+  return v;
+}
+
+double NekRSDriver::temperature_at(int32_t local_elem) const
+{
+  Expects(local_elem < n_local_elem());
+
+  double sum0 = 0.;
+  double sum1 = 0.;
+  for (int32_t i = 0; i < n_gll_; ++i) {
+    auto idx = local_elem * n_gll_ + i;
+    sum0 += rho_cp_[idx] * temperature_[idx];
+    sum1 += rho_cp_[idx];
+  }
+  return sum0 / sum1;
+}
+
+std::vector<double> NekRSDriver::temperature_local() const
+{
+  std::vector<double> t(n_local_elem());
+  for (int32_t i = 0; i < n_local_elem(); ++i) {
+    t[i] = this->temperature_at(i);
+  }
+
+  return t;
+}
+
+std::vector<double> NekRSDriver::density_local() const
+{
+  nekrs::copyToNek(time_, tstep_);
+  std::vector<double> local_densities(n_local_elem());
+
+  for (int32_t i = 0; i < n_local_elem(); ++i) {
+    if (this->in_fluid_at(i) == 1) {
+      auto T = this->temperature_at(i);
+      // nu1 returns specific volume in [m^3/kg]
+      local_densities[i] = 1.0e-3 / iapws::nu1(pressure_bc_, T);
+    } else {
+      local_densities[i] = 0.0;
+    }
+  }
+
+  return local_densities;
+}
+
+int NekRSDriver::in_fluid_at(int32_t local_elem) const
+{
+  // In NekRS, element_info_[i] == 1 if i is a *solid* element
+  Expects(local_elem < n_local_elem());
+  return element_info_[local_elem] == 1 ? 0 : 1;
+}
+
+std::vector<int> NekRSDriver::fluid_mask_local() const
+{
+  std::vector<int> mask(n_local_elem());
+  for (int32_t i = 0; i < n_local_elem(); ++i) {
+    mask[i] = in_fluid_at(i);
+  }
+  return mask;
+}
+
+int NekRSDriver::set_heat_source_at(int32_t local_elem, double heat)
+{
+  for (int i = 0; i < n_gll_; ++i) {
+    localq_->at(local_elem * n_gll_ + i) = heat;
+  }
+  return 0;
+}
+
+void NekRSDriver::open_lib_udf()
+{
+  lib_udf_handle_ = dlopen(lib_udf_name_.c_str(), RTLD_LAZY);
+  if (lib_udf_handle_ == NULL) {
+    throw std::runtime_error(std::string{"dlopen error for localq: "} + dlerror());
+  }
+  void* localq_void = dlsym(lib_udf_handle_, "localq");
+  if (dlerror() != NULL) {
```

Comment:
```suggestion
  if (dlerror()) {
```

Reply:
Fixed in https://github.com/enrico-dev/enrico/pull/113/commits/e3f457c7035dd31924c2caccd4a59c9eaf021f92